docs/ocaml/00-outline.md

```diff
@@ -0,0 +1,143 @@
+1. Intro
+  - Why Functional Programming?
+    - mutability
+    - encapsulation
+    - functional programming in Java is tedious (but possible)
+  - Goal: Introduce what we will build at the very end (after async)
+    - and show demo possible (or a snippet of code)
+
+2. Basics
+  - Utop (how to run)
+  - End an expression with `;;`
+  - Integers (and “all values have a type”)
+  - Strings (and concatenation)
+  - Lists (they’re linked lists)
+  - Tuples (they’re not lists)
+  - Let-bindings
+    - Shadowing
+  (exercise on using let-bindings)
+
+3. Functions
+    - function types (show `int_of_string`)
+    - function application
+    - functions are values (`let f = int_of_string ;;`)
+    - currying and two arg functions (`(``+``)` like 5)
+      - partial application
+      - redefine it on a tuple (but we usually use currying): this is different!
+    - Anonymous functions (`fun` keyword)
+    - Defining functions
+    - An example higher-order function (not map/filter/fold — nonrecursive)
+      - Maybe something with parsing
+  (examples and exercises)
+
+4. ADTs and patternmatching (no recursive functions)
+  - “enums” in Java
+  - “enums” in OCaml (are sum types)
+    - Int or String
+    - This is what you can do that you can’t do in Java
+    - Pattern-matching
+  - “Algebraic of datatypes” (but don’t describe these now)
+  - Recursive Datatypes
+    - Lists (of ints)
+    - Binary Tree (of ints)
+  (examples and exercises:
+    Should showcase: a problem that requires using creating new ADTs,
+        using pattern matching on them
+  )
+
+5. Recursive functions
+  - `let rec`
+  - `length` (simple) (on list of ints)
+    - Worth stepping through `length [1,2,3]`
+  - string-concatenation on a tree (basically a fold, but don’t say that)
+  - recursive function mental model
+    - Think from the bottom up (start from the end)
+      - Base case
+      - Recursive case
+  (examples and exercises:
+    Should showcase: the same problems that we’ll do again in higher-order
+    function section (example: write a function that removes the even numbers
+    from a list, later we’ll ask them again with filter, (do it for a map and a
+    fold also))
+  )
+
+6. Parametric Polymorphism (don’t use the words parametric polymorphism)
+  - Datatypes
+    - Lists (of ‘a)
+    - Trees (of ‘a)
+    - Option types
+      - no nulls, no billion dollar mistake
+    - Result type “either”
```

Comment:
We should compare parametric polymorphism with Java Generics.

Reply:
+1. Or templates.